Between/Not between expression

elpaso opened the pull request:
Support for SQL-like `BETWEEN` and `NOT BETWEEN` expressions.


Funded by: Kanton Solothurn

elpaso committed:
BETWEEN initial research

elpaso committed:
BETWEEN expression

TODO: type casting, SQL compiler, help text

elpaso committed:
SQL expression compiler and more tests for BETWEEN

elpaso committed:
Flake8

elpaso committed:
NOT BETWEEN help

elpaso committed:
NULL between tests

elpaso committed:
Fix function help

elpaso committed:
No braces

nirvn commented:
Very cool @elpaso ! Have you tested against date,datetime values?

elpaso commented:
> Very cool @elpaso ! Have you tested against date,datetime values?

Yes: https://github.com/qgis/QGIS/pull/47771/files#diff-78762f6fed0edd0a03a082ac901d682e7812bf3841a50ca205481f096d0f619fR2098

elpaso committed:
pointcloud exp node

nyalldawson commented on resources/function_help/json/BETWEEN:
```diff
@@ -0,0 +1,18 @@
+  }],
+  "examples": [{
+    "expression": "'B' BETWEEN 'A' AND 'C'",
+    "returns": "1"
```
Shouldn't these examples be returning true/false? (Not 0/1)

nyalldawson replied:
I believe that's just a historic remnant because those operators were introduced before we had proper Boolean support in fields/expressions

elpaso replied:
Should I change them all?

elpaso replied:
Maybe better as a follow up PR...

nyalldawson replied:
Yeah, because we should backport for the existing functions

nyalldawson commented on python/core/auto_generated/expression/qgsexpressionnodeimpl.sip.in:
```diff
@@ -248,6 +248,69 @@ Returns the index node, representing an array element index or map key.
 
+};
+
+class QgsExpressionNodeBetweenOperator: QgsExpressionNode
```
The sip casting code in QgsExpressionNode will need handling of this new class

nyalldawson commented on resources/function_help/json/BETWEEN:
```diff
@@ -0,0 +1,18 @@
+  "name": "BETWEEN",
+  "type": "operator",
+  "groups": ["Operators"],
+  "description": "Returns 1 if value is within the specified range. To test for exclusion `NOT BETWEEN` can be used.",
```
```suggestion
  "description": "Returns 1 if value is within the specified range. The range is considered inclusive of the bounds. To test for exclusion `NOT BETWEEN` can be used.",
```

nyalldawson commented on resources/function_help/json/NOT BETWEEN:
```diff
@@ -0,0 +1,18 @@
+  "name": "NOT BETWEEN",
+  "type": "operator",
+  "groups": ["Operators"],
+  "description": "Returns 1 if value is not within the specified range.",
```
```suggestion
  "description": "Returns 1 if value is not within the specified range. The range is considered inclusive of the bounds.",
```

nyalldawson commented on src/core/expression/qgsexpressionnodeimpl.cpp:
```diff
@@ -1793,6 +1793,128 @@ QList<const QgsExpressionNode *> QgsExpressionNodeInOperator::nodes() const
+{
+  QgsExpressionNodeBinaryOperator lowBound { QgsExpressionNodeBinaryOperator::BinaryOperator::boGE, mNode->clone(), mLowerBound->clone() };
+  QgsExpressionNodeBinaryOperator highBound { QgsExpressionNodeBinaryOperator::BinaryOperator::boLE, mNode->clone(), mHigherBound->clone() };
+  const QVariant lowBoundValue = lowBound.eval( parent, context );
+  const QVariant highBoundValue = highBound.eval( parent, context );
+  if ( lowBoundValue.isNull() || highBoundValue.isNull() )
```
```suggestion
  const QVariant lowBoundValue = lowBound.eval( parent, context );
   if ( lowBoundValue.isNull() )
   {
     return QVariant();
   }
  const QVariant highBoundValue = highBound.eval( parent, context );
  if ( highBoundValue.isNull() )
```

avoids potentially evaluating the high bound when we can shortcut early

nyalldawson replied:
actually you could also move the construction of highBound to after testing lowBoundValue too, that'll potentially save some work.

elpaso replied:
Actually, looking at the NULL handling in the oracle implementation https://oracle.su/docs/11g/timesten.112/e13070/ttsql269.htm I think I'll need to make a smarter implementation and do not always return NULL...

nyalldawson commented on src/core/expression/qgsexpressionnodeimpl.cpp:
```diff
@@ -1793,6 +1793,128 @@ QList<const QgsExpressionNode *> QgsExpressionNodeInOperator::nodes() const
+QVariant QgsExpressionNodeBetweenOperator::evalNode( QgsExpression *parent, const QgsExpressionContext *context )
+{
+  QgsExpressionNodeBinaryOperator lowBound { QgsExpressionNodeBinaryOperator::BinaryOperator::boGE, mNode->clone(), mLowerBound->clone() };
+  QgsExpressionNodeBinaryOperator highBound { QgsExpressionNodeBinaryOperator::BinaryOperator::boLE, mNode->clone(), mHigherBound->clone() };
```
doesn't this approach mean we are evaluating mNode twice?

nyalldawson replied:
I think you could avoid this by evaluating mNode upfront, and then using a static value node in lowBound/highBound instead.

elpaso replied:
I refactored the evaluation code trying to optimize but also to handle NULLs when it is possible to return a boolean, it is a bit more complex now but I made sure that all code paths are covered by tests.

nyalldawson commented on src/core/qgssqlexpressioncompiler.cpp:
```diff
@@ -307,6 +307,13 @@ QgsSqlExpressionCompiler::Result QgsSqlExpressionCompiler::compileNode( const Qg
+    case QgsExpressionNode::ntBetweenOperator:
+    {
+      const QgsExpressionNodeBetweenOperator *n = static_cast<const QgsExpressionNodeBetweenOperator *>( node );
+      result = n->dump();
```
Are you confident that all subclasses of QgsSqlExpressionCompiler support between? (i.e. sqlite, hana, sql server, postgres, spatialite, oracle)

nyalldawson replied:
We'll need a test added to feature source test base to cover this -- somewhere here: https://github.com/qgis/QGIS/blob/master/tests/src/python/featuresourcetestbase.py#L161

elpaso replied:
I'm pretty confident that BETWEEN is widely supported by all SQL dialects, I checked postgres and OGR but I'll add a test as you recommend.

elpaso replied:
@nyalldawson I'm making progresses with the iterator tests but there is a wierd issue with datetime on spatialite provider (works fine when connecting through OGR though, IIRC there was some logic in OGR to handle this Zulu issue):

the expression is compiled as
`select * from somedata where dt BETWEEN '2020-05-03T12:13:14Z' AND '2020-05-04T12:14:14Z'`

but this does not match the test data `2020-05-03T12:13:14` (note: no `Z`ulu), column type for `dt` is `TIMESTAMP`, any idea?

nyalldawson replied:
Hmm, not sure sorry. I do know that sqlite datetime handling is completely horrible though. If you're struggling with getting this working well I'd suggest just blocking the compilation of this node type for the spatialite provider (better slow and safe vs fast but inaccurate)

elpaso replied:
I could skip the between operator, but this seems to be a more general issue: this works

`select * from somedata where dt = '2020-05-03T12:13:14'`

this doesn't

`select * from somedata where dt = '2020-05-03T12:13:14Z'`

It seems like the `Z` comes from here:
https://github.com/qgis/QGIS/blob/master/src/core/qgssqliteexpressioncompiler.cpp#L130

The docs here say `Z` is a no-op https://www.sqlite.org/lang_datefunc.html but it doesn't seem to be the case.

Try to remove `Z` now ...

nyalldawson replied:
> Try to remove Z now ...

Oh, please **don't** do that. It's likely to break the most important use case of date time compilation we have -- which is to ensure that datetime filters on gpkg work correctly and take advantage of attribute indices. I'd FAR rather "between" is never compiled for sqlite sources then see that regress. 

Ping @nirvn, who may have further insights on sqlite/gpkg date handling

elpaso replied:
Yeah, it was only a test, but the problem is not limited to BETWEEN as you can see from the `=` sample query above.

elpaso replied:
Here is a full session to demonstrate the issue:

```sql
@blackhole ~/dev/QGIS (between-expression)$ sqlite3 /home/dev/QGIS/tests/testdata/provider/spatialite.db
SQLite version 3.31.1 2020-01-27 19:55:54
Enter ".help" for usage hints.
sqlite> select cnt, dt from somedata; 
100|2020-05-03T12:13:14
200|2020-05-04T12:14:14
300|
400|2021-05-04T13:13:14
-200|2020-05-04T12:13:14
sqlite> select cnt, dt from somedata where dt = '2020-05-03T12:13:14Z';
sqlite> select cnt, dt from somedata where dt = '2020-05-03T12:13:14';
100|2020-05-03T12:13:14
```

elpaso committed:
Address PR review comments and handle NULLS

Refactor the evaluation code, handle NULLs to
return a boolean when it is possible.

elpaso committed:
Some more tests for compiler

nyalldawson commented:
New changes are looking good @elpaso , thanks!

elpaso committed:
Remove Z from make_datetime SQLITE compiler

elpaso committed:
Disable BETWEEN for HANA

elpaso commented:
@mrylov test is failing for HANA, I'm trying to disable BETWEEN altogether for HANA now unless you now how to fix it, note that I cannot/don't want to test HANA locally.

elpaso committed:
NoUnaryMinus for OGR, no idea why

nyalldawson commented on tests/src/python/test_provider_shapefile.py:
```diff
@@ -218,11 +218,14 @@ def uncompiledFilters(self):
                        'to_time("time") = to_time(\'000www14ww13ww12www\',\'zzzwwwsswwmmwwhhwww\')',
                        'to_datetime("dt", \'yyyy-MM-dd hh:mm:ss\') + make_interval(days:=1) <= make_datetime(2020, 5, 4, 12, 13, 14)',
-                       'to_datetime("dt", \'yyyy-MM-dd hh:mm:ss\') + make_interval(days:=0.01) <= make_datetime(2020, 5, 4, 12, 13, 14)'
+                       'to_datetime("dt", \'yyyy-MM-dd hh:mm:ss\') + make_interval(days:=0.01) <= make_datetime(2020, 5, 4, 12, 1NoUnaryMinus3, 14)',
```
```suggestion
                       'to_datetime("dt", \'yyyy-MM-dd hh:mm:ss\') + make_interval(days:=0.01) <= make_datetime(2020, 5, 4, 12, 13, 14)',
```

I assume this was an accidental paste?

elpaso committed:
Update tests/src/python/test_provider_shapefile.py

Co-authored-by: Nyall Dawson <nyall.dawson@gmail.com>

elpaso committed:
Revert "Remove Z from make_datetime SQLITE compiler"

This reverts commit 0909682a3d01e54081c623220a49c62d8f725d3b.

elpaso committed:
Do not compile make_datetime for spatialite

Because of the Zulu issue:

sqlite3 /home/dev/QGIS/tests/testdata/provider/spatialite.db
SQLite version 3.31.1 2020-01-27 19:55:54
Enter ".help" for usage hints.
sqlite> select cnt, dt from somedata;
100|2020-05-03T12:13:14
200|2020-05-04T12:14:14
300|
400|2021-05-04T13:13:14
-200|2020-05-04T12:13:14
sqlite> select cnt, dt from somedata where dt = '2020-05-03T12:13:14Z';
sqlite> select cnt, dt from somedata where dt = '2020-05-03T12:13:14';
100|2020-05-03T12:13:14

elpaso commented:
@nyalldawson I've reverted th `Z` change and blocked `make_datetime` SQL compilation for spatialite only, we should be good to go now.

nyalldawson reviewed (approved):
LGTM!
